core/tx_list.go

```diff
@@ -248,25 +249,62 @@ func (l *txList) Overlaps(tx *types.Transaction) bool {
 //
 // If the new transaction is accepted into the list, the lists' cost and gas
 // thresholds are also potentially updated.
-func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Transaction) {
+func (l *txList) Add(tx *types.Transaction, priceBump uint64, baseFee *big.Int) (bool, *types.Transaction) {
 	// If there's an older better transaction, abort
+	eip1559 := false
 	old := l.txs.Get(tx.Nonce())
 	if old != nil {
-		threshold := new(big.Int).Div(new(big.Int).Mul(old.GasPrice(), big.NewInt(100+int64(priceBump))), big.NewInt(100))
 		// Have to ensure that the new gas price is higher than the old gas
 		// price as well as checking the percentage threshold to ensure that
 		// this is accurate for low (Wei-level) gas price replacements
-		if old.GasPrice().Cmp(tx.GasPrice()) >= 0 || threshold.Cmp(tx.GasPrice()) > 0 {
-			return false, nil
+		if old.GasPrice() != nil {
+			threshold := new(big.Int).Div(new(big.Int).Mul(old.GasPrice(), big.NewInt(100+int64(priceBump))), big.NewInt(100))
+			if tx.GasPrice() != nil {
+				if old.GasPrice().Cmp(tx.GasPrice()) >= 0 || threshold.Cmp(tx.GasPrice()) > 0 {
+					return false, nil
+				}
+			} else {
+				eip1559 = true
+				newGasPrice := new(big.Int).Add(baseFee, tx.GasPremium())
+				if newGasPrice.Cmp(tx.FeeCap()) > 0 {
+					newGasPrice.Set(tx.FeeCap())
+				}
+				if old.GasPrice().Cmp(newGasPrice) >= 0 || threshold.Cmp(newGasPrice) > 0 {
+					return false, nil
+				}
+			}
```

Comment:
```suggestion
			if  tx.GasPrice() != nil && old.GasPrice().Cmp(tx.GasPrice()) >= 0 || threshold.Cmp(tx.GasPrice()) > 0 {
				return false, nil
			}
			eip1559 = true
			newGasPrice := new(big.Int).Add(baseFee, tx.GasPremium())
			if newGasPrice.Cmp(tx.FeeCap()) > 0 {
				newGasPrice.Set(tx.FeeCap())
			}
			if old.GasPrice().Cmp(newGasPrice) >= 0 || threshold.Cmp(newGasPrice) > 0 {
				return false, nil
			}
```
Lots of nested here so trying to clean it up. Would this work?

Reply:
Thanks! Made all these suggested changes except this one as it's not quite equivalent.

Reply:
Also agree that the code could be broken up much better, apologies. If I find the time I will do that. In the original repo there are issues which follow the progress on specific sections/issues. It's not perfect since the process was somewhat iterative, but that does provide some more structure to the commits: https://github.com/vulcanize/go-ethereum-EIP1559/pulls?q=is%3Apr+is%3Aclosed